deploy/tests/awstools/test_awstools.py

```diff
@@ -0,0 +1,79 @@
+class TestSNS(object):
+    """Test functions in awstools that work with SNS"""
+
+    # When we're testing awstools.get_snsname_arn(), we don't also want to be
+    # testing awstools.aws_resource_names(). mainly because we're subjugating
+    # the boto3.client factory (in the second test below) and also because we want to be able to run these
+    # tests while not on a AWS EC2 host and aws_resource_names() assumes that it
+    # can lookup the AWS instance-id via 169.254.169.254 API
+    @patch('awstools.awstools.aws_resource_names', return_value={'snsname': 'Testing'})
+    # register the moto backend for sns
+    @mock_sns
+    def test_get_snsname_arn_sanity(self, aws_res_mock):
+        """Simple call returns properly formed ARN"""
+        # local imports of code-under-test ensure moto has mocks
+        # registered before any possible calls out to AWS
+        from awstools.awstools import get_snsname_arn
+
+        arn = get_snsname_arn()
+
+        client = boto3.client('sns')
+        response = client.get_topic_attributes(TopicArn=arn)
+
+        # output will normally be captured and suppressed but printed
+        # iff the test fails. So, leaving in something that dumps the response
+        # can be useful. See https://docs.pytest.org/en/4.6.x/capture.html
+        pprint(response)
+        assert response['ResponseMetadata']['HTTPStatusCode'] == 200
+        assert response['Attributes']['TopicArn'] == arn
+
+        # check that our mock of aws_resource_names was used
+        aws_res_mock.assert_called_once()
+
+
+    @patch('awstools.awstools.aws_resource_names', return_value={'snsname': 'Testing'})
+    @mock_sns
+    def test_get_snsname_arn_auth_exception_handling(self, aws_res_mock):
+        """AuthorizationError is intercepted and re-raised as AssertionError"""
+        # local imports of code-under-test ensure moto has mocks
+        # registered before any possible calls out to AWS
+        from awstools.awstools import get_snsname_arn
+
+        # create a mock SNS client that returns what we tell it to
+        client = boto3.client('sns')
+        stub = Stubber(client)
+        stub.add_client_error('create_topic', service_error_code='AuthorizationError')
+        stub.activate()
+
+
+        # since firesim manager code doesn't take clients as method parameters
+        # now we mock boto3.client to return our stubbed client
+        with patch.object(boto3._get_default_session(), 'client', return_value=client) as mock_session:
+            with pytest.raises(AssertionError):
+                get_snsname_arn()
+
+            # TODO we could mock rootLogger.critical to capture it's calls and args and validate that we're seeing the correct "nice" message
+
+            # make sure get_snsname_arn() actually called out to get a sns
+            # client, otherwise we aren't testing what we think we are
+            mock_session.assert_called_once_with('sns')
+
+        aws_res_mock.assert_called_once()
```

Comment:
I dislike this pattern of using `@patch` as a decorator because it is unclear that the `@patch` decorator is providing the parameter, especially when mixed with other decorators and given that pytest provides fixtures as parameters.  

There is a `pytest-mock` plugin to pytest that thinly wraps `mock` so that it's usage is a little cleaner in pytest.  if we use that plugin, this code would look like this
```python
    @mock_sns
    def test_get_snsname_arn_auth_exception_handling(self, mocker):
        """AuthorizationError is intercepted and re-raised as AssertionError"""
        # local imports of code-under-test ensure moto has mocks
        # registered before any possible calls out to AWS
        from awstools.awstools import get_snsname_arn
        
        mocker.patch('awstools.awstools.aws_resource_names', return_value={'snsname': 'Testing'})

        # create a mock SNS client that returns what we tell it to
        client = boto3.client('sns')
        stub = Stubber(client)
        stub.add_client_error('create_topic', service_error_code='AuthorizationError')
        stub.activate()


        # since firesim manager code doesn't take clients as method parameters
        # now we mock boto3.client to return our stubbed client
        mocked_client_factory = patch.object(boto3._get_default_session(), 'client', return_value=client)
        with pytest.raises(AssertionError):
              get_snsname_arn()

        # TODO we could mock rootLogger.critical to capture it's calls and args and validate that we're seeing the correct "nice" message

        # make sure get_snsname_arn() actually called out to get a sns
        # client, otherwise we aren't testing what we think we are
        mocked_client_factory.assert_called_once_with('sns')

        awstools.awstools.aws_resource_names.assert_called_once()
```

Or, we could do one of the alternatives discussed in https://github.com/pytest-dev/pytest-mock/blob/3a4b13b9413e9c67b91e55b972698da1241e3ee2/README.rst#why-bother-with-a-plugin 

I do think that the pytest-mock plugin makes the test code cleaner.  I didn't add it because:
* I was trying to limit the number of libraries I added 
* I was lazy and didn't want to spend time figuring out which version would work with the version of mock that we have installed.